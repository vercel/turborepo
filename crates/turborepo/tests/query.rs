mod common;

#[test]
fn test_query() -> Result<(), anyhow::Error> {
    check_json!(
        "basic_monorepo",
        "npm@10.5.0",
        "query",
        "get package that doesn't exist" => "query { package(name: \"doesnotexist\") { path } }",
    );

    Ok(())
}

#[cfg(not(windows))]
#[test]
fn test_double_symlink() -> Result<(), anyhow::Error> {
    check_json!(
        "oxc_repro",
        "npm@10.5.0",
        "query",
        "get_dependencies" => "query {
                 file(path: \"./index.js\") {
                    path
                    dependencies {
                      files { items { path } }
                      errors { items { message import } }
                    }
                 }
              }",
    );
    Ok(())
}

#[test]
fn test_ast() -> Result<(), anyhow::Error> {
    // Separate because the `\\` -> `/` filter isn't compatible with ast
    check_json!(
        "turbo_trace",
        "npm@10.5.0",
        "query",
        "get `main.ts` with ast" => "query { file(path: \"main.ts\") { path ast } }",
    );

    Ok(())
}

#[test]
fn test_trace() -> Result<(), anyhow::Error> {
    insta::with_settings!({ filters => vec![(r"\\\\", "/")]}, {
        check_json!(
            "turbo_trace",
            "npm@10.5.0",
            "query",
            "get `main.ts`" => "query { file(path: \"main.ts\") { path } }",
            "get `main.ts` with dependencies" => "query { file(path: \"main.ts\") { path, dependencies { files { items { path } } } } }",
            "get `button.tsx` with dependencies" => "query { file(path: \"button.tsx\") { path, dependencies { files { items { path } } } } }",
            "get `circular.ts` with dependencies" => "query { file(path: \"circular.ts\") { path dependencies { files { items { path } } } } }",
            "get `invalid.ts` with dependencies" => "query { file(path: \"invalid.ts\") { path dependencies { files { items { path } } errors { items { import } } } } }",
            "get `main.ts` with depth = 0" => "query { file(path: \"main.ts\") { path dependencies(depth: 1) { files { items { path } } } } }",
            "get `with_prefix.ts` with dependencies" => "query { file(path: \"with_prefix.ts\") { path dependencies { files { items { path } } } } }",
            "get `import_value_and_type.ts` with all dependencies" => "query { file(path: \"import_value_and_type.ts\") { path dependencies(importType: ALL) { files { items { path } } } } }",
            "get `import_value_and_type.ts` with type dependencies" => "query { file(path: \"import_value_and_type.ts\") { path dependencies(importType: TYPES) { files { items { path } } } } }",
            "get `import_value_and_type.ts` with value dependencies" => "query { file(path: \"import_value_and_type.ts\") { path dependencies(importType: VALUES) { files { items { path } } } } }",
            "get `export_conditions` with dependencies" => "query { file(path: \"export_conditions.js\") { path dependencies(depth: 1) { files { items { path } } } } }",
        );

        Ok(())
    })
}

#[test]
fn test_trace_on_monorepo() -> Result<(), anyhow::Error> {
    insta::with_settings!({ filters => vec![(r"\\\\", "/")]}, {
        check_json!(
            "turbo_trace_monorepo",
            "npm@10.5.0",
            "query",
            "get `apps/my-app/index.ts` with dependencies" => "query { file(path: \"apps/my-app/index.ts\") { path dependencies { files { items { path } } errors { items { message } } } } }",
            "get `packages/utils/index.ts` with dependents" => "query { file(path: \"packages/utils/index.ts\") { path dependents { files { items { path } } errors { items { message } } } } }",
            "get `packages/another/index.js` with dependents" => "query { file(path: \"packages/another/index.js\") { path dependents { files { items { path } } errors { items { message } } } } }",
        );

        Ok(())
    })
}

#[test]
fn test_reverse_trace() -> Result<(), anyhow::Error> {
    check_json!(
        "turbo_trace",
        "npm@10.5.0",
        "query",
        "get `button.tsx` with dependents" => "query { file(path: \"button.tsx\") { path dependents { files { items { path } } } } }",
        "get `link.tsx` with all dependents" => "query { file(path: \"link.tsx\") { path dependents(importType: ALL) { files { items { path } } } } }",
        "get `link.tsx` with type dependents" => "query { file(path: \"link.tsx\") { path dependents(importType: TYPES) { files { items { path } } } } }",
        "get `link.tsx` with value dependents" => "query { file(path: \"link.tsx\") { path dependents(importType: VALUES) { files { items { path } } } } }",
    );

    Ok(())
}
