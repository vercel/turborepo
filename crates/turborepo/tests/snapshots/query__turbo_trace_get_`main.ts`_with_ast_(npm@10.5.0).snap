---
source: crates/turborepo/tests/query.rs
expression: query_output
---
{
  "data": {
    "file": {
      "path": "main.ts",
      "ast": {
        "type": "Module",
        "span": {
          "start": 1,
          "end": 173
        },
        "body": [
          {
            "type": "ImportDeclaration",
            "span": {
              "start": 1,
              "end": 39
            },
            "specifiers": [
              {
                "type": "ImportSpecifier",
                "span": {
                  "start": 10,
                  "end": 16
                },
                "local": {
                  "type": "Identifier",
                  "span": {
                    "start": 10,
                    "end": 16
                  },
                  "ctxt": 0,
                  "value": "Button",
                  "optional": false
                },
                "imported": null,
                "isTypeOnly": false
              }
            ],
            "source": {
              "type": "StringLiteral",
              "span": {
                "start": 24,
                "end": 38
              },
              "value": "./button.tsx",
              "raw": "\"./button.tsx\""
            },
            "typeOnly": false,
            "with": null,
            "phase": "evaluation"
          },
          {
            "type": "ImportDeclaration",
            "span": {
              "start": 40,
              "end": 64
            },
            "specifiers": [
              {
                "type": "ImportDefaultSpecifier",
                "span": {
                  "start": 47,
                  "end": 50
                },
                "local": {
                  "type": "Identifier",
                  "span": {
                    "start": 47,
                    "end": 50
                  },
                  "ctxt": 0,
                  "value": "foo",
                  "optional": false
                }
              }
            ],
            "source": {
              "type": "StringLiteral",
              "span": {
                "start": 56,
                "end": 63
              },
              "value": "./foo",
              "raw": "\"./foo\""
            },
            "typeOnly": false,
            "with": null,
            "phase": "evaluation"
          },
          {
            "type": "ImportDeclaration",
            "span": {
              "start": 65,
              "end": 100
            },
            "specifiers": [
              {
                "type": "ImportDefaultSpecifier",
                "span": {
                  "start": 72,
                  "end": 78
                },
                "local": {
                  "type": "Identifier",
                  "span": {
                    "start": 72,
                    "end": 78
                  },
                  "ctxt": 0,
                  "value": "repeat",
                  "optional": false
                }
              }
            ],
            "source": {
              "type": "StringLiteral",
              "span": {
                "start": 84,
                "end": 99
              },
              "value": "repeat-string",
              "raw": "\"repeat-string\""
            },
            "typeOnly": false,
            "with": null,
            "phase": "evaluation"
          },
          {
            "type": "VariableDeclaration",
            "span": {
              "start": 102,
              "end": 130
            },
            "ctxt": 0,
            "kind": "const",
            "declare": false,
            "declarations": [
              {
                "type": "VariableDeclarator",
                "span": {
                  "start": 108,
                  "end": 129
                },
                "id": {
                  "type": "Identifier",
                  "span": {
                    "start": 108,
                    "end": 114
                  },
                  "ctxt": 0,
                  "value": "button",
                  "optional": false,
                  "typeAnnotation": null
                },
                "init": {
                  "type": "NewExpression",
                  "span": {
                    "start": 117,
                    "end": 129
                  },
                  "ctxt": 0,
                  "callee": {
                    "type": "Identifier",
                    "span": {
                      "start": 121,
                      "end": 127
                    },
                    "ctxt": 0,
                    "value": "Button",
                    "optional": false
                  },
                  "arguments": [],
                  "typeArguments": null
                },
                "definite": false
              }
            ]
          },
          {
            "type": "ExpressionStatement",
            "span": {
              "start": 132,
              "end": 148
            },
            "expression": {
              "type": "CallExpression",
              "span": {
                "start": 132,
                "end": 147
              },
              "ctxt": 0,
              "callee": {
                "type": "MemberExpression",
                "span": {
                  "start": 132,
                  "end": 145
                },
                "object": {
                  "type": "Identifier",
                  "span": {
                    "start": 132,
                    "end": 138
                  },
                  "ctxt": 0,
                  "value": "button",
                  "optional": false
                },
                "property": {
                  "type": "Identifier",
                  "span": {
                    "start": 139,
                    "end": 145
                  },
                  "value": "render"
                }
              },
              "arguments": [],
              "typeArguments": null
            }
          },
          {
            "type": "ExpressionStatement",
            "span": {
              "start": 149,
              "end": 166
            },
            "expression": {
              "type": "CallExpression",
              "span": {
                "start": 149,
                "end": 165
              },
              "ctxt": 0,
              "callee": {
                "type": "Identifier",
                "span": {
                  "start": 149,
                  "end": 155
                },
                "ctxt": 0,
                "value": "repeat",
                "optional": false
              },
              "arguments": [
                {
                  "spread": null,
                  "expression": {
                    "type": "StringLiteral",
                    "span": {
                      "start": 156,
                      "end": 161
                    },
                    "value": "foo",
                    "raw": "\"foo\""
                  }
                },
                {
                  "spread": null,
                  "expression": {
                    "type": "NumericLiteral",
                    "span": {
                      "start": 163,
                      "end": 164
                    },
                    "value": 5.0,
                    "raw": "5"
                  }
                }
              ],
              "typeArguments": null
            }
          },
          {
            "type": "ExpressionStatement",
            "span": {
              "start": 167,
              "end": 173
            },
            "expression": {
              "type": "CallExpression",
              "span": {
                "start": 167,
                "end": 172
              },
              "ctxt": 0,
              "callee": {
                "type": "Identifier",
                "span": {
                  "start": 167,
                  "end": 170
                },
                "ctxt": 0,
                "value": "foo",
                "optional": false
              },
              "arguments": [],
              "typeArguments": null
            }
          }
        ],
        "interpreter": null
      }
    }
  }
}
